src/persist-client/src/read.rs

```diff
@@ -142,41 +209,89 @@ where
         timeout: Duration,
     ) -> Result<Vec<ListenEvent<K, V, T, D>>, LocationError> {
         trace!("Listen::poll_next timeout={:?}", timeout);
+        let deadline = Instant::now() + timeout;
+
+        let (batch_keys, desc) = self
+            .machine
+            .next_listen_batch(deadline, &self.frontier)
+            .await?;
+        let updates = self.fetch_batch(deadline, &batch_keys).await?;
+        let mut ret = Vec::with_capacity(2);
+        if !updates.is_empty() {
+            ret.push(ListenEvent::Updates(updates));
+        }
+        ret.push(ListenEvent::Progress(desc.upper().clone()));
+        self.frontier = desc.upper().clone();
+        return Ok(ret);
+    }
+
+    /// Test helper to read from the listener until the given frontier is
+    /// reached.
+    #[cfg(test)]
+    pub async fn read_until(
+        &mut self,
+        ts: &T,
+    ) -> Result<Vec<ListenEvent<K, V, T, D>>, LocationError> {
+        use crate::NO_TIMEOUT;
+
         let mut ret = Vec::new();
-        loop {
-            let state = self.shard.clone().into_inner();
-            let state = state.lock().await;
-            let new_upper = Antichain::from(
-                state
-                    .upper
-                    .iter()
-                    .map(|x| T::decode(*x))
-                    .collect::<Vec<_>>(),
-            );
-            if PartialOrder::less_than(&self.upper, &new_upper) {
-                // TODO: We could order contents to avoid scanning the entire
-                // thing every time, but this impl is meant as a placeholder, so
-                // let's see how far we get without that.
-                let mut updates = Vec::new();
-                for (k, v, t, d) in state.contents.iter() {
-                    let t = T::decode(*t);
-                    if self.as_of.less_than(&t) && self.upper.less_equal(&t) {
-                        debug_assert_eq!(new_upper.less_equal(&t), false);
-                        let k = K::decode(&k);
-                        let v = V::decode(&v);
-                        let d = D::decode(*d);
-                        updates.push(((k, v), t, d));
+        while self.frontier.less_than(ts) {
+            let mut next = self.poll_next(NO_TIMEOUT).await?;
+            ret.append(&mut next);
+        }
+        return Ok(ret);
+    }
+
+    async fn fetch_batch(
+        &self,
+        deadline: Instant,
+        keys: &[String],
+    ) -> Result<Vec<((Result<K, String>, Result<V, String>), T, D)>, LocationError> {
+        let mut ret = Vec::new();
+        for key in keys {
+            // TODO: Deduplicate this with the logic in SnapshotIter.
+            let value = loop {
+                let value = self.blob.get(deadline, &key).await?;
+                match value {
+                    Some(x) => break x,
+                    // If the underlying impl of blob isn't linearizable, then we
```

Comment:
realization I had while typing this up: if we keep the seqno capabilities on readhandle, this logic is super straightforward even when BlobMulti isn't linearizable. just retry until it shows up. but if we remove them and have to start re-fetching more recent versions of the state as we iterate a snapshot, then a non-linearizable BlobMulti becomes far more complicated to reason about